cpp/dolfin/fem/FormIntegrals.cpp

```diff
@@ -21,68 +24,143 @@ FormIntegrals::FormIntegrals()
-
-  if (_default_cell_integral)
-    _tabulate_tensor_cell.push_back(_default_cell_integral->tabulate_tensor);
-
-  // -- Create exterior facet integrals
-  std::unique_ptr<ufc_exterior_facet_integral, decltype(&std::free)>
-      _default_exterior_facet_integral{
-          ufc_form.create_exterior_facet_integral(-1), &std::free};
+  std::vector<int> interior_facet_integral_ids(
+      ufc_form.num_interior_facet_integrals);
+  ufc_form.get_interior_facet_integral_ids(interior_facet_integral_ids.data());
+  for (const auto& id : interior_facet_integral_ids)
+  {
+    ufc_interior_facet_integral* interior_facet_integral
+        = ufc_form.create_interior_facet_integral(id);
+    assert(interior_facet_integral);
+    register_tabulate_tensor_interior_facet(
+        id, interior_facet_integral->tabulate_tensor);
+    std::free(interior_facet_integral);
+  }
 
-  if (_default_exterior_facet_integral)
+  // Not currently working
+  std::vector<int> vertex_integral_ids(ufc_form.num_vertex_integrals);
+  ufc_form.get_vertex_integral_ids(vertex_integral_ids.data());
+  if (vertex_integral_ids.size() > 0)
   {
-    // Extract tabulate tensor function
-    _tabulate_tensor_exterior_facet.push_back(
-        _default_exterior_facet_integral->tabulate_tensor);
+    throw std::runtime_error(
+        "Vertex integrals not supported. Under development.");
   }
 }
 //-----------------------------------------------------------------------------
 const std::function<void(PetscScalar*, const PetscScalar*, const double*, int)>&
-FormIntegrals::get_tabulate_tensor_fn_cell(int i) const
+FormIntegrals::get_tabulate_tensor_fn_cell(unsigned int i) const
 {
+  if (i > _tabulate_tensor_cell.size())
+    throw std::runtime_error("Invalid integral index");
+
   return _tabulate_tensor_cell[i];
 }
 //-----------------------------------------------------------------------------
 const std::function<void(PetscScalar*, const PetscScalar*, const double*, int,
                          int)>&
-FormIntegrals::get_tabulate_tensor_fn_exterior_facet(int i) const
+FormIntegrals::get_tabulate_tensor_fn_exterior_facet(unsigned int i) const
 {
+  if (i > _tabulate_tensor_exterior_facet.size())
+    throw std::runtime_error("Invalid integral index");
+
   return _tabulate_tensor_exterior_facet[i];
 }
 //-----------------------------------------------------------------------------
-void FormIntegrals::set_tabulate_tensor_cell(
+const std::function<void(PetscScalar*, const PetscScalar* w, const double*,
+                         const double*, int, int, int, int)>&
+FormIntegrals::get_tabulate_tensor_fn_interior_facet(unsigned int i) const
+{
+  if (i > _tabulate_tensor_interior_facet.size())
+    throw std::runtime_error("Invalid integral index");
+
+  return _tabulate_tensor_interior_facet[i];
+}
+//-----------------------------------------------------------------------------
+int _insert_ids(std::vector<int>& ids, int new_id)
```

Comment:
We are avoiding this design (modification of input arguments) in DOLFIN-X. Add extra return argument for `ids`.

Reply:
This is a short local helper function, not a public interface. I'd rather cut and paste 3 times than add more return args.